civisocial.php

```diff
@@ -203,15 +203,24 @@ function civisocial_civicrm_buildForm($formName, &$form) {
       ));
 
       // Populate fields
-      // @todo: Come up with a more general solution
       $defaults = array();
-      $defaults['email-5'] = $oAuthUser['email'];
-      $defaults['email-Primary'] = $oAuthUser['email'];
-      $defaults['first_name'] = $oAuthUser['first_name'];
-      $defaults['billing_first_name'] = $oAuthUser['first_name'];
-      if ($oAuthUser['last_name']) {
-        $defaults['last_name'] = $oAuthUser['last_name'];
-        $defaults['billing_last_name'] = $oAuthUser['last_name'];
+      foreach ($form->_elements as $element) {
+        if ('text' == $element->_type) {
+          $elementName = $element->_attributes['name'];
+
+          if (preg_match('/first_name/', $elementName)) {
+            $defaults[$elementName] = $oAuthUser['first_name'];
+          }
+          elseif (preg_match('/last_name/', $elementName)) {
+            $defaults[$elementName] = $oAuthUser['last_name'];
+          }
+          elseif (preg_match('/^name$/', $elementName)) {
+            $defaults[$elementName] = $oAuthUser['name'];
+          }
+          elseif (preg_match('/email/', $elementName)) {
+            $defaults[$elementName] = $oAuthUser['email'];
+          }
+        }
       }
```

Comment:
How about doing like this

$formFields = array(
    'first_name',
    'last_name',
    'email',
  );
  $elements = array_keys($form->_elementIndex);
  foreach ($formFields as $formField) {
    $matches  = preg_grep ("/{$formField}/", $elements);
    foreach ($matches as $elementName) {
      $defaults[$elementName] = $oAuthUser[$formField];
    }   
  }